packages/sol-compiler/src/compiler.ts

```diff
@@ -107,97 +131,117 @@ export class Compiler {
-            solcVersion = semver.maxSatisfying(availableCompilerVersions, solcVersionRange);
-        }
-        const fullSolcVersion = binPaths[solcVersion];
-        const compilerBinFilename = path.join(SOLC_BIN_DIR, fullSolcVersion);
-        let solcjs: string;
-        const isCompilerAvailableLocally = fs.existsSync(compilerBinFilename);
-        if (isCompilerAvailableLocally) {
-            solcjs = fs.readFileSync(compilerBinFilename).toString();
-        } else {
-            logUtils.log(`Downloading ${fullSolcVersion}...`);
-            const url = `${constants.BASE_COMPILER_URL}${fullSolcVersion}`;
-            const response = await fetchAsync(url);
-            const SUCCESS_STATUS = 200;
-            if (response.status !== SUCCESS_STATUS) {
-                throw new Error(`Failed to load ${fullSolcVersion}`);
+    private async _compileContractsAsync(contractNames: string[]): Promise<void> {
+        const inputsByVersion: {
+            [solcVersion: string]: {
+                standardInput: solc.StandardInput;
+                contractsToCompile: string[];
+            };
+        } = {};
+
+        const contractData: {
+            [contractPath: string]: {
+                currentArtifactIfExists: ContractArtifact | void;
+                sourceTreeHashHex: string;
+                contractName: string;
+            };
+        } = {};
+
+        for (const contractName of contractNames) {
+            const contractSource = this._resolver.resolve(contractName);
+            contractData[contractSource.path] = {
+                contractName,
+                currentArtifactIfExists: await getContractArtifactIfExistsAsync(this._artifactsDir, contractName),
+                sourceTreeHashHex: `0x${this._getSourceTreeHash(
+                    path.join(this._contractsDir, contractSource.path),
+                ).toString('hex')}`,
+            };
+            let shouldCompile = false;
+            if (_.isUndefined(contractData[contractSource.path].currentArtifactIfExists)) {
+                shouldCompile = true;
+            } else {
+                const currentArtifact = contractData[contractSource.path].currentArtifactIfExists as ContractArtifact;
+                const isUserOnLatestVersion = currentArtifact.schemaVersion === constants.LATEST_ARTIFACT_VERSION;
+                const didCompilerSettingsChange = !_.isEqual(currentArtifact.compiler.settings, this._compilerSettings);
+                const didSourceChange =
+                    currentArtifact.sourceTreeHashHex !== contractData[contractSource.path].sourceTreeHashHex;
+                shouldCompile = !isUserOnLatestVersion || didCompilerSettingsChange || didSourceChange;
+            }
+            if (!shouldCompile) {
+                continue;
             }
-            solcjs = await response.text();
-            fs.writeFileSync(compilerBinFilename, solcjs);
+            let solcVersion = this._solcVersionIfExists;
```

Comment:
I addressed this concern by taking a slightly different approach, making it so that `solcVersion` *can't* be undefined.  That's the intent of this variable anyways (note that right after the orignal assignment there was a check to set it if it was undefined).  Let me know what you think.